python-checks/src/main/java/org/sonar/python/checks/WeakSSLProtocolCheck.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.python.PythonSubscriptionCheck;
+import org.sonar.python.api.tree.PyNameTree;
+import org.sonar.python.api.tree.Tree;
+import org.sonar.python.semantic.Symbol;
+
+@Rule(key = "S4423")
+public class WeakSSLProtocolCheck extends PythonSubscriptionCheck {
+  private static final List<String> WEAK_PROTOCOL_CONSTANTS = Arrays.asList(
+    "ssl.PROTOCOL_SSLv2",
+    "ssl.PROTOCOL_SSLv3",
+    "ssl.PROTOCOL_SSLv23",
+    "ssl.PROTOCOL_TLS",
+    "ssl.PROTOCOL_TLSv1",
+    "ssl.PROTOCOL_TLSv1_1",
+    "OpenSSL.SSL.SSLv2_METHOD",
+    "OpenSSL.SSL.SSLv3_METHOD",
+    "OpenSSL.SSL.SSLv23_METHOD",
+    "OpenSSL.SSL.TLSv1_METHOD",
+    "OpenSSL.SSL.TLSv1_1_METHOD"
+  );
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.NAME, ctx -> {
+      PyNameTree pyNameTree = (PyNameTree) ctx.syntaxNode();
+      Symbol symbol = ctx.symbolTable().getSymbol(pyNameTree);
+      if (isWeakProtocol(pyNameTree, symbol)) {
+        ctx.addIssue(pyNameTree, "Change this code to use a stronger protocol.");
+      }
+    });
+  }
+
+  private static boolean isWeakProtocol(PyNameTree pyNameTree, @Nullable Symbol symbol) {
+    Predicate<String> matchWeakProtocol;
+    if (symbol == null) {
+      matchWeakProtocol = s -> s.substring(s.lastIndexOf('.') + 1).equals(pyNameTree.name());
```

Comment:
I don't get why we need that. I think we have a problem because, in the unit test, this method returns true only when we go through this branch. The other branch, where we use the symbol, doesn't seem to raise issues. Maybe it's related to https://github.com/SonarSource/sonar-python/blob/master/python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java#L329

Maybe we should either:
* create a ticket to fix the problem later
* or fix the problem in the symbol table right away

Reply:
Problem is deeper : this is completely linked to semantic:  this is about solving dotted names : in order to get fully qualified name of the constant we need to figure out fqn of its parent and right now the SymbolTableVisitor does not deal with this kind of constructs (if i'm correct, it only deals with the first part of a dotted name). 
As I did not want to address semantic in the implementation of this new rule and as I managed to make it work I believe this is ok. 
Not sure which kind of ticket I should create for such case to be honest.

Reply:
Let me jump in the discussion. Just to clarify, currently the semantic has some limitations and inconsistencies regarding symbol resolution:

```python
from foo import A
from foo.A import fn

A.fn() # Here "getSymbol()" will retrieve the correct symbol for "A.fn" PythonGrammar.ATTRIBUTE_REF (corresponding to PyQualifiedExpressionTree), but not for "fn" PythonGrammar.ATOM node

fn() # Here "getSymbol()" will retrieve the correct symbol for "fn" PythonGrammar.ATOM node
```

I agree we shouldn't fix the semantics right now, I can create a ticket for this.

Reply:
@andrea-guarino-sonarsource be my guest and please link it here !